### PR TITLE
perf(python): Avoid dispatching `Series.head/tail` to the expression engine

### LINKED ISSUE #12928: Series `head` / `tail` is about 60+ times slower than DataFrame ones
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
from timeit import timeit

df = pl.DataFrame({"a": pl.arange(0, 10000, eager=True), "c": 1, "b": "a", "x": 42.42})
df_series = df["a"]
plain_series = pl.Series("x", pl.arange(0, 10000, eager=True))

def go(code):
    print(code, timeit(code, number=100_000, globals=globals()))

go("df.head(0)")
go("df.head(100)")
go("df.tail(0)")
go("df.tail(100)")
go("df_series.head(0)")
go("df_series.head(100)")
go("df_series.tail(0)")
go("df_series.tail(100)")
go("plain_series.head(0)")
go("plain_series.head(100)")
go("plain_series.tail(0)")
go("plain_series.tail(100)")
```

### Log output

```shell
df.head(0) 0.18811989994719625
df.head(100) 0.1828387000132352
df.tail(0) 0.12071079993620515
df.tail(100) 0.10931550001259893
df_series.head(0) 7.409043599967845
df_series.head(100) 8.439290300011635
df_series.tail(0) 11.03056729992386
df_series.tail(100) 10.96223199996166
plain_series.head(0) 9.089440800016746
plain_series.head(100) 9.058857599971816
plain_series.tail(0) 10.974754099966958
plain_series.tail(100) 10.8858651999617
```

### Issue description

Doing head/tail on a series is almost a couple of magnitude slower than doing the same operation on a dataframe.

### Expected behavior

These operations have a comparable time than dataframe. Since they are doing less work I would expect them to be faster, but comparable time is good enough.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.19.19
Index type:           UInt32
Platform:             Windows-10-10.0.19045-SP0
Python:               3.10.13 | packaged by Anaconda, Inc. | (main, Sep 11 2023, 13:24:38) [MSC v.1916 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          2.2.1
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
matplotlib:           3.8.0
numpy:                1.26.2
openpyxl:             3.1.2
pandas:               2.1.3
pyarrow:              14.0.1
pydantic:             1.10.13
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           3.1.9
```

</details>

### PR DESCRIPTION
**Pull request by mcrumiller**
Resolves #12928.

### CODE REVIEW

File: py-polars/src/series/mod.rs [hunk 1]
```diff
@@ -679,6 +679,14 @@ impl PySeries {
         self.series.clear().into()
     }
 
+    fn head(&self, n: Option<usize>) -> Self {
+        self.series.head(n).into()
+    }
```
Comment (stinodego) [inline py-polars/src/series/mod.rs:684]: The input here is not an Option type (same for tail):

```suggestion
    fn head(&self, n: usize) -> Self {
        self.series.head(Some(n)).into()
    }
```

**Review by stinodego — CHANGES_REQUESTED** [review]
I wonder why dispatching to DataFrame is so much slower? :thinking:

**Comment by mcrumiller** [comment]
> I wonder why dispatching to DataFrame is so much slower? 🤔

From my research it's due to `select` taking up all of the time. If we want to solve this issue for other cases, it might be worth having a behind-the-scenes fast path for single-column selection by name into an ldf.

**Review by stinodego — APPROVED** [review]
Right, we'll have to put some thought into the dispatching. Especially when it comes to often-used/cheap methods.

**Comment by CaselIT** [comment]
> I wonder why dispatching to DataFrame is so much slower? 🤔

Doing the more naive version seems to be very fast
(on a different pc)
```py
go("plain_series.head(0)")
go("plain_series.head(100)")
go("plain_series.tail(0)")
go("plain_series.tail(100)")
go("plain_series.to_frame().head(0).to_series()")
go("plain_series.to_frame().head(100).to_series()")
go("plain_series.to_frame().tail(0).to_series()")
go("plain_series.to_frame().tail(100).to_series()")
```
```
plain_series.head(0) 4.969168399926275
plain_series.head(100) 5.323634900152683
plain_series.tail(0) 5.622998500009999
plain_series.tail(100) 5.60176460002549
plain_series.to_frame().head(0).to_series() 0.14708300004713237
plain_series.to_frame().head(100).to_series() 0.15118449996225536
plain_series.to_frame().tail(0).to_series() 0.1474038001615554
plain_series.to_frame().tail(100).to_series() 0.14568820013664663
```

**Comment by mcrumiller** [comment]
I've added direct dispatch to a lot more Series functions, will update momentarily.

**Comment by stinodego** [comment]
> I've added direct dispatch to a lot more Series functions, will update momentarily.

That's a bit premature. I'd like to discuss this with the team first.

**Comment by mcrumiller** [comment]
Hi @stinodego, no problem. I had assumed that any argument for direct dispatch to `head`/`tail` could also be applied to any arbitrary function, but I can see why we wouldn't want to add code bloat for minor performance improvements. For the record, here is what I have in my commit (but have not pushed):

* `head`
* `tail`
* `any`
* `all`
* `product`
* `nan_max`
* `nan_min`
* `std`
* `var`
* `cut`
* `qcut`
* `value_counts`
* `entropy`
* `is_between`
* `slice`

**Comment by mcrumiller** [comment]
I'll push for now so it's easier to gauge the impact, we can always revert anything we need.

**Comment by ritchie46** [comment]
@mcrumiller can you revert this PR to the head and tail then we can merge this? You can open an issue for the other ones.

**Review by stinodego — APPROVED** [review]

**Pull request merged**